src/apps/weblib/ts-api/data.ts

```diff
@@ -130,6 +160,13 @@ export class Data {
 		return true
 	}
 
+	private _convertDimension(values: string[]): { categories: string[], indexes: number[] } {
+		const uniques = new Set(values)
+		const categories = Array.from(uniques)
+		const indexes = values.map((value) => categories.indexOf(value))
```

Comment:
indexOf has linear time complexity, which could cause bad performance here it there are a lot of categories.
I suggest to use a Map instead of the Set with the categories as keys and the indexes as values. Something like this:
```javascript
    const uniques = new Map();
    const categories = [];
    const indexes = new Array(values.length);
    values.forEach((value, index) => {
        if (!uniques.has(value)) {
            uniques.set(value, categories.length);
            categories.push(value);
        }
        indexes[index] = uniques.get(value);
    });
    return { categories, indexes };
```

Reply:
probably a 
```js
let got = uniques.get(); 
if (got === undefined) {
   ...
   got = categories.length;
}
... = got;
```
is better to eliminate the unnecessary `.has()` call.

Reply:
seems right